kotlinlib/src/mill/kotlinlib/js/KotlinJSModule.scala

```diff
@@ -0,0 +1,509 @@
+      callMain = callMain(),
+      moduleKind = moduleKind(),
+      produceSourceMaps = kotlinJSSourceMap(),
+      sourceMapEmbedSourcesKind = kotlinJSSourceMapEmbedSources(),
+      sourceMapNamesPolicy = kotlinJSSourceMapNamesPolicy(),
+      splitPerModule = splitPerModule(),
+      esTarget = kotlinJSESTarget(),
+      kotlinVersion = kotlinVersion(),
+      destinationRoot = T.dest,
+      extraKotlinArgs = kotlincOptions(),
+      worker = kotlinWorkerTask()
+    )
+  }
+
+  // endregion
+
+  // region private
+
+  private def createKlib(destFolder: os.Path, irPathRef: PathRef): PathRef = {
+    val outputPath = destFolder / s"${moduleName()}.klib"
+    Jvm.createJar(
+      outputPath,
+      Agg(irPathRef.path),
+      mill.api.JarManifest.MillDefault,
+      fileFilter = (_, _) => true
+    )
+    PathRef(outputPath)
+  }
+
+  private[kotlinlib] def kotlinJsCompile(
+      outputMode: OutputMode,
+      allKotlinSourceFiles: Seq[PathRef],
+      irClasspath: Option[PathRef],
+      librariesClasspath: Agg[PathRef],
+      callMain: Boolean,
+      moduleKind: ModuleKind,
+      produceSourceMaps: Boolean,
+      sourceMapEmbedSourcesKind: SourceMapEmbedSourcesKind,
+      sourceMapNamesPolicy: SourceMapNamesPolicy,
+      splitPerModule: Boolean,
+      esTarget: Option[String],
+      kotlinVersion: String,
+      destinationRoot: os.Path,
+      extraKotlinArgs: Seq[String],
+      worker: KotlinWorker
+  )(implicit ctx: mill.api.Ctx): Result[CompilationResult] = {
+    val kotlinVersionSegments = kotlinVersion.split("\\.")
+    if (
+      kotlinVersionSegments(0).toInt == 2 || (
+        kotlinVersionSegments(0).toInt == 1
+          && (kotlinVersionSegments(1).toInt <= 7
+            || (kotlinVersionSegments(1).toInt == 8 && kotlinVersionSegments(2).toInt < 20))
+      )
+    ) {
+      // have to put this restriction, because for older versions some compiler options either didn't exist or
+      // had different names. It is possible to go to the lower version supported with a certain effort.
+      ctx.log.error("Minimum supported Kotlin version for JS target is 1.8.20, maximum is 1.9.25")
+      return Result.Aborted
+    }
+
+    val inputFiles = irClasspath match {
+      case Some(x) => Seq(s"-Xinclude=${x.path.toIO.getAbsolutePath}")
+      case None => allKotlinSourceFiles.map(_.path.toIO.getAbsolutePath)
+    }
+
+    // TODO: Cannot support Kotlin 2+, because it doesn't publish .jar anymore, but .klib files only. Coursier is not
+    //  able to work with that (unlike Gradle, which can leverage .module metadata).
+    // https://repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-stdlib-js/2.0.20/
+    val librariesCp = librariesClasspath.map(_.path)
+      .filter(os.exists)
+      .filter(isKotlinJsLibrary)
+
+    val innerCompilerArgs = Seq.newBuilder[String]
+    // classpath
+    innerCompilerArgs ++= Seq("-libraries", librariesCp.iterator.mkString(File.pathSeparator))
```

Comment:
Is there any documentation page for all these flags that we can link in a comment?

Reply:
Sure, added. In fact there is an [official reference](https://kotlinlang.org/docs/compiler-reference.html#kotlin-js-compiler-options), but to discover the full set of options it is better to look at the [code](https://github.com/JetBrains/kotlin/blob/v1.9.25/compiler/cli/cli-common/src/org/jetbrains/kotlin/cli/common/arguments/K2JSCompilerArguments.kt).